Unable to use union types in query parameters definition

## Sorting

- **I'm submitting a ...**

  - [ ] bug report
  - [x] feature request
  - [ ] support request

- I confirm that I
  - [x] used the [search](https://github.com/lukeautry/tsoa/search?type=Issues) to make sure that a similar issue hasn't already been submit

## Expected Behavior

I expected to be able to define a set of query parameters using a type that is a union of multiple types. For example:


```ts
interface FiltersA { name: string, priceFrom: number, priceTo: number }
interface FiltersB { position: string, contractType: string }

interface AdsFilterParams = FiltersA | FiltersB

@Get("/list/{categorySlug}")
public async getAds(
  @Path() categorySlug: string,
  @Queries() filters: AdsFilterParams,
): Promise<API.ResourceCollectionResponse<AdData>> {
  // ...
}

```

## Current Behavior
However, when attempting to generate routes with this configuration, the route generator throws an error:

```
Generate routes error.
 GenerateMetadataError: @Queries('filters') only support 'refObject' or 'nestedObjectLiteral' types. If you want only one query parameter, please use the '@Query' decorator.
```

## Possible Solution

It would be greatly beneficial if TSOA could support union types in query parameter definitions. This enhancement would enable developers to define more granular and category-specific interfaces, leading to cleaner and more maintainable code.

## Context (Environment)

Version of the library: 5.1.1
Version of NodeJS: 19.7.0

Update:  I tracked tihs down, @Query is able to handle union, but not with "undefined".

```
union {
  dataType: 'union',
  types: [ { dataType: 'string' }, { dataType: 'undefined' } ]
}

```

If I change it to 
`@Query() lang: ISOLanguageCode = req.locale?.language || 'de'`
it works,
funny part now too, with this TSOA can handle it, but typescript complains:
`@Query() lang?: ISOLanguageCode = req.locale?.language`

error TS1015: Parameter cannot have question mark and initializer.

What about if we skip the `undefineds` in tsoa parsing?
https://github.com/lukeautry/tsoa/blob/6380e4861a7ef802c18cf3455db9c1fb2200de5a/packages/cli/src/metadataGeneration/parameterGenerator.ts#L442

`return !parameterType.types.map(t => t.dataType === 'undefined' ? true : this.supportPathDataType(t)).some(t => t === false);
`
as this only changes the check if it is valid, the output of the swagger file looks still right:
```
 -
                    in: query
                    name: lang
                    required: false
                    schema:
                        $ref: '#/components/schemas/ISOLanguageCode'
                -

```

--- a/packages/cli/src/metadataGeneration/parameterGenerator.ts
+++ b/packages/cli/src/metadataGeneration/parameterGenerator.ts
@@ -440,7 +440,7 @@ export class ParameterGenerator {
 
     if (parameterType.dataType === 'union') {
       // skip undefined inside unions
-      return !parameterType.types.map(t => t.dataType === 'undefined' ? true : this.supportPathDataType(t)).some(t => t === false); 
+      return !parameterType.types.map(t => t.dataType === 'undefined' || this.supportPathDataType(t)).some(t => t === false);
     }
 
     return false;

--- a/packages/cli/src/metadataGeneration/parameterGenerator.ts
+++ b/packages/cli/src/metadataGeneration/parameterGenerator.ts
@@ -439,7 +439,8 @@ export class ParameterGenerator {
     }
 
     if (parameterType.dataType === 'union') {
-      return !parameterType.types.map(t => this.supportPathDataType(t)).some(t => t === false);
+      // skip undefined inside unions
+      return !parameterType.types.map(t => t.dataType === 'undefined' ? true : this.supportPathDataType(t)).some(t => t === false); 
     }
 
     return false;

--- a/tests/fixtures/controllers/getController.ts
+++ b/tests/fixtures/controllers/getController.ts
@@ -327,6 +327,11 @@ export class GetTestController extends Controller {
   public async getIndexedValueGeneric(): Promise<IndexedValueGeneric<IndexedValueTypeReference>> {
     return 'FOO';
   }
+
+  @Get('UnionTypeWithDefault')
+  public async getUnionTypeWithDefault(@Query() unionType: 'a' | 'b' | undefined = 'a'): Promise<void> {
+    return;
+  }
 }
 
 export interface ErrorResponse {

